[DPE-4645] Cruise Control Roles implementation

Batalex opened the pull request:
Supersede #199, #205, #206

TODO before properly opening for review:
- [x] Clean up all remnants of the external balancer relation (mostly in models.py)
- [x] Document changes (especially the 'dependent' thing in the various lib handlers so that we can decide to keep it or not)
- [x] Add more tests (unit for coverage, integration for running CC)
- [x] ~Add custom events in broker role for better control flow~
- [x] Fix failing integration tests
- [x] Add docstrings everywhere


## Changelog and design decisions

- **New pytest markers**: `broker`, `balancer` to facilitate the local development feedback loop
- **Broker logic moved to `src/events/broker.py`**. With **roles** plural, we needed to refactor things a little bit to separate the broker logic from the charm's.
- **New peer relation `balancer`** used to share balancer-specific data and trigger events accordingly.
- **New admin user `"balancer"`** created upon relating to zookeeper
- **Enforced types in unit test**. This proved to be immensely helpful when moving stuff around


### To be discussed
- **Reduced mandatory storage from 10G to 1G**: this comes from when we had **role** singular and did not want to allocate 10G to CC. IMO we can keep 1G as it will be easier on the CI runners 
- Moving the broker logic to a new file had some consequences: first, not all handlers could be removed from the top-level logic (like the restart one), but we now have to use two different workloads. Handlers were modified to reflect this change with a `dependent` argument instead of passing the charm instance.
- We wait until we have 3 brokers to start CC because we write the necessary topics on 3 replicas. Is that a sensible default?
- ~Specific .env for each workload: do we keep them, are they useful?~

marcoppenheimer commented on src/events/broker.py:
```diff
@@ -0,0 +1,328 @@
+            self._on_config_changed(event)
+
+    def _on_storage_detaching(self, _: StorageDetachingEvent) -> None:
+        """Handler for `storage_detaching` events."""
+        # in the case where there may be replication recovery may be possible
+        if self.charm.state.brokers and len(self.charm.state.brokers) > 1:
+            self.charm._set_status(Status.REMOVED_STORAGE)
+        else:
+            self.charm._set_status(Status.REMOVED_STORAGE_NO_REPL)
+
+        self.charm.on.config_changed.emit()
```
**todo:** I believe we need to update the `storages` here for the unit, is that right?

Batalex replied:
I wonder about updating here because if the unit is being removed, it does not matter if the unit itself declares 0 storage or if the leader removes the unit from the `broker-capabilities`. If anything else, the current behavior maybe preserves from a wrong `capacityJBOD.json`, if 0 storage for a broker is not allowed. But you are right: if there are several storages on the unit, we do not update the databag at the moment.


Gave it a shot anyway in 89cd5191633bef7b93900e1a5eab24f1b5a1f73f, if that causes issues I'll figure out something else

marcoppenheimer commented on src/managers/config.py:
```diff
@@ -97,20 +107,12 @@ def advertised_listener(self) -> str:
 
-class ConfigManager:
-    """Manager for handling Kafka configuration."""
+class _ConfigManager:
```
**nitpicky suggestion:** I didn't love the private class. Maybe change it to `SharedConfigManager`? `BaseConfigManager`? `CommonConfigManager`?

marcoppenheimer replied:
**praise:** After reading through the rest of the file though, I think this choice for inheriting the base config is a really neat solution!

Batalex replied:
Yep, on second thought, I do not think it makes that much sense in the first place to have private elements in a charm. This might be a hot take since we use private methods for event handlers (e.g. `_on_update_status`), but a charm does not expose a public API in the first place.
Anyway, sorry about the rambling, renamed in e92b42bd795ee2ae006f570f4c604bee877b437c

marcoppenheimer commented on src/managers/config.py:
```diff
@@ -465,22 +507,13 @@ def set_environment(self) -> None:
 
         updated_env = current_env | map_env(updated_env_list)
         content = "\n".join([f"{key}={value}" for key, value in updated_env.items()])
         self.workload.write(content=content, path="/etc/environment")
+        self.workload.write(content=content, path=f'{BROKER.paths["CONF"]}/.env')
```
**question:** I know we talked about this, but what was the reasoning behind having two environments with the same content here?

Batalex replied:
So, the issue at first was that I was using a Jaas config file to pass credentials. Since the binary either looks in a hardcoded path or through the `KAFKA_OPTS` env var, I wanted a way to have different env vars depending on the workload. At first, I was just overwriting `/etc/environment` before every command, and wanted an easy way to access those vars when SSh-ing into the machine. Hence, the .env files.

Then, I thought that I might as well avoid `/etc/environment` altogether and figure out a way to use those specific .env. This is doable any time we use `subprocess`functions ourselves, but not when we use stuff from `operator_libs_linux`, so we were still stuck with overwriting `/etc/environment`.

In the end, we do not use a Jaas file any more, so the var issue is gone too. I can get rid of the .env files, but I wanted to check if there is still a use for them

marcoppenheimer replied:
I think long term, we should make a PR to `operator_libs_linux.v1.snap` lib upstream to support passing env, and then the `.env` file idea is MUCH better than a polluted `/etc/environment`, but let's do that at a later date when we have some spare time.

Batalex replied:
.envs removed in  ad5961c16b72616404a171ddf114ada5c8c45e5f

marcoppenheimer commented on src/workload.py:
```diff
@@ -155,15 +148,37 @@ def get_service_pid(self) -> int:
         for pid in java_processes.split():
             with open(f"/proc/{pid}/cgroup", "r") as fid:
                 content = "".join(fid.readlines())
+                logger.error(content)
+                logger.error(f"{self.SNAP_NAME}.{self.service}")
```
**todo:** If we still want to keep this (i.e it wasn't just debugging), let's wrap these together in to a single message.

Batalex replied:
Done in 3809a7ee84bf52180b1ba6c0559c265bc2559d90

marcoppenheimer commented on tests/integration/test_balancer.py:
```diff
@@ -0,0 +1,73 @@
+pytestmark = pytest.mark.balancer
+
+
+def balancer_is_running(model_full_name: str | None) -> bool:
+    try:
+        check_output(
+            f"JUJU_MODEL={model_full_name} juju ssh kafka/0 sudo -i 'curl http://localhost:9090/kafkacruisecontrol/state'",
+            stderr=PIPE,
+            shell=True,
+            universal_newlines=True,
+        )
+    except CalledProcessError:
+        return False
+    return True
```
**suggestion:** I think it's fine to just raise `CalledProcessError` here. If the test fails, we'd want to get the actual traceback, rather than it being hidden in the `AssertionError` that would have been raised.

We could do:
```python
assert balancer_is_running()
```

Batalex replied:
Yes, that makes perfect sense. I first wrote this helper having the `assert not balancer_is_running()` statement in mind, but I fully agree that keeping the traceback is way better. Changed in 97f9a44f678c30e812875aa4bca0dac2a80e6b9d

marcoppenheimer commented on tests/integration/test_balancer.py:
**praise:** These are honestly great tests! I think we need to cover more test cases, but that can be done later. 

Off the top of my head, I'd like to see a check that ensures that leader removal re-starts CC on the new leader, for example.

marcoppenheimer reviewed (commented):
This is fantastic. I left a LOT of comments (it's a big PR!), but genuinely, for your first big contribution, you've done a lot right.
You've picked up the style and idiosyncrasies really fast, there were loads of examples where I thought 'neat trick!' without it being too clever, and I also saw loads of places where you made small minor improvements unrelated to the feature.

Great work, really impressive stuff :).

I think the only thing I'd want to see more of (later, not for this PR) are more tests to cover the branching behavior, and balancer specific tests, but as this is a feature branch, it's OK to not cover that for now.

Batalex commented on src/events/broker.py:
```diff
@@ -0,0 +1,328 @@
+                }
+            )
+
+        if (
+            not self.upgrade.idle
+            and not self.charm.state.cluster.internal_user_credentials
+            and self.charm.state.zookeeper.zookeeper_connected
+        ):
+            # Create credentials for missing internal user, to reconcile state during upgrades
+            self.zookeeper._on_zookeeper_changed(event)  # type: ignore
```
Yes, it's a new addition that I added to make the upgrade integration test pass. The gist of it is: since we now have a new `balancer` internal user, we need to find a way to create it when we update the charm from the latest `stable` or `edge`. Otherwise, the application would get stuck on NO_BROKER_CREDS.

I don't really like it, though. I would prefer any kind of reconciliation logic to be in `upgrade.py`

marcoppenheimer replied:
I think we should put this in the `events/upgrade.py` section, something like:
```python
def apply_backwards_compatibility_fixes(self) -> None:
        """A range of functions needed for backwards compatibility."""
        # Rev.113 - needed for allowing log-level upgrades (DPE-123)
        self.charm.config_manager.set_server_jvmflags()
```

Batalex commented on src/charm.py:
```diff
@@ -111,274 +69,92 @@ def __init__(self, *args):
+
+        if (
+            BALANCER.value not in self.config.roles
+            and self.unit.is_leader()
```
Yes, you are absolutely right. This additional condition is there because I originally added loggers, and it was polluting the logs with unnecessary "Cruise control stopped" messages for units where it was not running in the first place.
I do not remember why I removed them, though. Do you prefer:
- adding the loggers back, and keeping the condition?
- adding the loggers back, and removing the condition?
- only removing the condition?

Batalex commented on src/charm.py:
```diff
@@ -111,274 +69,92 @@ def __init__(self, *args):
 
-    def _restart(self, event: EventBase) -> None:
-        """Handler for `rolling_ops` restart events."""
+        The RollingOpsManager expecting a charm instance, we cannot move this method to the broker logic.
+        """
```
If I understand your question correctly, yes, we could, but we would need to have `_disable_enable_restart_broker` in charm.py anyway because `callback_override` is a str parameter. So I figured we might as well keep the restart logic in this file

Batalex commented on src/core/cluster.py:
```diff
@@ -186,6 +215,15 @@ def port(self) -> int:
             else SECURITY_PROTOCOL_PORTS["SASL_PLAINTEXT"].client
         )
 
+    @property
+    def internal_port(self) -> int:
+        """Return the port to be used for an internal client."""
```
Is your version a strict 1:1 match with this one? This might be worth unifying before merging

marcoppenheimer replied:
Dw about it, my feature is coming later, I'll work off yours.

Batalex commented on src/core/cluster.py:
```diff
@@ -213,7 +263,7 @@ def planned_units(self) -> int:
 
     @property
-    def ready_to_start(self) -> Status:
+    def ready_to_start(self) -> Status:  # noqa: C901
```
Because of all the if statements, the cyclomatic complexity of this method is above our quality threshold. I silenced [this specific rule](https://docs.astral.sh/ruff/rules/#mccabe-c90) here, but we could rework the entire thing to not have any ifs. We could use something like [itertools.compress](https://docs.python.org/3/library/itertools.html#itertools.compress).

What do you think? I do not mind at all, but I was a little worried that this was a "too clever one-liner for our own good" solution

Batalex commented on src/core/models.py:
```diff
@@ -454,3 +469,43 @@ def extra_user_roles(self) -> str:
+        super().__init__(relation, data_interface, None, substrate)
+        self.data_interface = data_interface
+
+    @property
+    def rack_aware(self) -> bool:
+        """Is the Kafka deployment rack aware?"""
+        if not self.relation:
+            return False
+
+        return (
+            self.data_interface.fetch_my_relation_field(
+                relation_id=self.relation.id, field="rack-aware"
+            )
+            == "true"
+            or False
+        )
```
Maybe I was doing something wrong with the data interface because I would get the `"false"` str which evaluates to _true-ish_.
Any ideas where things went wrong?

marcoppenheimer replied:
I think you can just remove the `or False` L497, which should still evaluate as a boolean.

Batalex replied:
Removed in  671643ebf594523a7a8d621551d69706d88f10d8, sorry I did not understand the first time

Batalex commented on src/core/structured_config.py:
```diff
@@ -235,3 +239,14 @@ def log_level_values(cls, value: str) -> str | None:
+    @classmethod
+    def roles_values(cls, value: str):
+        """Check roles values."""
+        roles = set(map(str.strip, value.split(",")))
```
Sometimes, one-liners are good 🤝 I was thinking of returning `sorted(roles)` instead of `list(roles)`, just in case we wanted some guarantee on the roles order. Any thought?

zmraul replied:
**nit:** I would expect the logic to follow `if role in roles_value: do something`, so sorting is not needed. Also, roles should never change within the context of a hook execution, so a `set` is the more "appropriate" type here I think.

Batalex commented on src/events/broker.py:
```diff
@@ -0,0 +1,328 @@
+            return
+
+        self.health = KafkaHealth(self)
+
```
Triggered me as well. Removed as part of 20ac8c27fdedd976d86fc0afd54763a7ab261130

Batalex commented on src/events/balancer.py:
```diff
@@ -0,0 +1,113 @@
+                    logger.info(f"Created topic {topic}")
+                except CalledProcessError:
+                    event.defer()
+                    return
```
Agreed, moved in 20ac8c27fdedd976d86fc0afd54763a7ab261130

Batalex commented on src/core/cluster.py:
```diff
@@ -238,4 +288,15 @@ def ready_to_start(self) -> Status:
             return Status.NO_BROKER_CREDS
 
+        # Additional checks specific to balancer role
+        if BALANCER.value in self.roles:
```
That's a great suggestion! Added in f637342b3704e98e72a26151409c157f58eb4864

Batalex commented on src/charm.py:
```diff
@@ -111,274 +69,92 @@ def __init__(self, *args):
+        getattr(logger, log_level.lower())(status.message)
+        self.unit.status = status
+
+    def _on_config_changed(self, _):
```
Alrighty, changed in 66b5f677aa8ce9cad5f9f4dc1b5bcd4ed8a86b68!

Batalex commented on src/events/broker.py:
```diff
@@ -0,0 +1,328 @@
+            self.charm.state.balancer.update(
+                {
+                    "broker-capacities": self.config_manager.broker_capacities,
+                    "rack_aware": json.dumps(bool(self.config_manager.rack_properties)),
```
You are absolutely right, this is an error since the relation field we fetch does use the hyphen. The only reason it was not caught until now is that we are not doing anything with it. Changed in 9bef8c5a8e377ae088a0304a93192b094c40e056

Batalex commented on src/events/balancer.py:
```diff
@@ -0,0 +1,113 @@
+
+    def _on_leader_elected(self, event: LeaderElectedEvent) -> None:
+        """Handler for `leader-elected` event."""
+        if not self.workload.active():
+            event.defer()
+            return
```
Sorry about that, the method is never called, and I forgot to remove it. I'll push the commit after I validate the "updating databag on detaching storage"  one

Batalex replied:
Removed method in d51f0498dcbbde6146c43befc93232ba3341cb68

marcoppenheimer commented on src/managers/config.py:
```diff
@@ -186,6 +188,32 @@ def heap_opts(self) -> str:
 
         return f"KAFKA_HEAP_OPTS='{' '.join(opts)}'"
 
+    @property
+    def security_protocol(self) -> AuthMechanism:
+        """Infers current charm security.protocol based on current relations."""
+        # FIXME: When we have multiple auth_mechanims/listeners, remove this method
+        return (
+            "SASL_SSL"
+            if (self.state.cluster.tls_enabled and self.state.unit_broker.certificate)
+            else "SASL_PLAINTEXT"
+        )
```
This should be moved to `state`, and just infer the ports from the security.protocol state here.

zmraul commented on src/events/balancer.py:
```diff
@@ -0,0 +1,117 @@
+        self.workload = BalancerWorkload()
+
+        # Fast exit after workload instantiation, but before any event observer
+        if BALANCER.value not in self.charm.config.roles:
+            return
```
**question:** I think this is a fairly sane way to do it. What about moving it to `charm.py` so it doesn't even need to reach this point?

Batalex replied:
I wanted `charm.py` to be completely unaware of what sub operators are running so that we don't have to use defensive programming everywhere.
Please let me know if I answered your question correctly

deusebio commented:
I don't think I will have time to review this today unfortunately. But even if I review, Alex, won't be able to address the comments. Feel free to merge this into the feature branch. I don't think that my approval is too important given it is on the feature branch. 

I'm also happy if you guys wants to merge the PR with failing tests, to then hand off development to Marc for the next pulse. So, overall, up to you what to do with the PR. I'll probably still have a look at the code here (so that I don't have to review everything in one go, and also provide general feedbacks early, probably providing some thoughts to Marc/Raul next week) but then I'll do the proper review once the feature has fully landed on the feature branch.

zmraul reviewed (commented):
Huge amazing work!

Regarding your points on the description:
> We wait until we have 3 brokers to start CC because we write the necessary topics on 3 replicas. Is that a sensible default?

To allow 1 unit deployment with CC we would need to change the `something.replication.factor.something` properties on the config files for CC with some dynamic value like so: `min([3, self.state.planned_units])`. In any case, I think enforcing 3 brokers is the right thing to do here. CC doesn't seem too useful with a single node deployment.

> Moving the broker logic to a new file had some consequences: first, not all handlers could be removed from the top-level logic (like the restart one), but we now have to use two different workloads. Handlers were modified to reflect this change with a dependent argument instead of passing the charm instance.

Having two workload seems to still fit the initial pattern of the refactor nicely. Though the workload could still be initialized at top level on `charm.py` and used through `self.charm.balancer_workload`.

Batalex commented:
Well, I think I addressed most of the comments on this PR. I hate to leave something in red for someone else to figure out, but I have not been able to make the upgrade test pass since 89cd5191633bef7b93900e1a5eab24f1b5a1f73f and I am now running out of time. 
Alex out :leaves:
